packages/compass-components/src/hooks/use-stacked-component.tsx

```diff
@@ -0,0 +1,94 @@
+import type { ComponentType, ForwardedRef } from 'react';
+import React, { createContext, useContext, useMemo } from 'react';
+import { cx, css } from '@leafygreen-ui/emotion';
+
+type StackedComponentProviderProps = {
+  zIndex?: number;
+};
+
+const StackedComponentContext =
+  createContext<StackedComponentProviderProps | null>(null);
+
+export const StackedComponentProvider = ({
+  zIndex,
+  children,
+}: StackedComponentProviderProps & { children: React.ReactNode }) => {
+  return (
+    <StackedComponentContext.Provider value={{ zIndex }}>
+      {children}
+    </StackedComponentContext.Provider>
+  );
+};
+
+export const useStackedComponent = () => {
+  return useContext(StackedComponentContext);
+};
+
+type StackedComponentProps<T extends boolean> = T extends true
+  ? { popoverZIndex?: number }
+  : T extends false
+  ? { className?: string }
+  : Record<string, never>;
+
+// TODO: LG-4109. This should be eventually supported by the LG design system
```

Comment:
Nit, it's not enforced, but we usually follow this format for todos

```suggestion
// TODO(LG-4109): This should be eventually supported by the LG design system
```